lib/logger/lib/logger.ex

```diff
@@ -447,53 +447,50 @@ defmodule Logger do
   @compile {:inline, __metadata__: 0}
 
   defp __metadata__ do
-    Process.get(@metadata) || {true, []}
+    Process.get(@metadata, true)
   end
 
   @doc """
   Alters the current process metadata according the given keyword list.
-
   This function will merge the given keyword list into the existing metadata,
   with the exception of setting a key to `nil`, which will remove that key
   from the metadata.
   """
   @spec metadata(metadata) :: :ok
   def metadata(keyword) do
-    {enabled?, metadata} = __metadata__()
-    Process.put(@metadata, {enabled?, into_metadata(keyword, metadata)})
-    :ok
-  end
+    case :logger.get_process_metadata() do
+      :undefined ->
+        :ok = :logger.set_process_metadata(filter_out_nils(keyword))
 
-  defp into_metadata([], metadata), do: metadata
-  defp into_metadata(keyword, metadata), do: into_metadata(keyword, [], metadata)
-
-  defp into_metadata([{key, nil} | keyword], prepend, metadata) do
-    into_metadata(keyword, prepend, :lists.keydelete(key, 1, metadata))
-  end
+      map when is_map(map) ->
+        merged = Enum.into(keyword, map)
+        metadata = filter_out_nils(merged)
```

Comment:
Is the intention to also filter the existing metadata or just the new ones? If only the new one, we could possibly merge & filter in one go
```elixir
for {_k, v} = elem <- keyword, v != nil, into: map, do: elem
```

Reply:
@michalmuskala when sending data to the backend all nils will be removed anyway, so there is not much difference.